scheduler/planner.go

```diff
@@ -0,0 +1,348 @@
+		}
+
+		totalPriority += t.Priority
+		expectedRuntime += t.FetchExpectedDuration()
+		numDeps += int64(t.NumDependents)
+	}
+
+	num := int64(len(unit.tasks))
+	priority := 1 + (totalPriority / num)
+
+	if inCommitQueue {
+		priority += 100
+	}
+
+	if inPatch {
+		unit.cachedValue += unit.distro.GetPatchZipperFactor()
+	}
+
+	unit.cachedValue += num
+	unit.cachedValue += priority
+	unit.cachedValue += priority * (numDeps / num)
+	unit.cachedValue += priority * unit.distro.GetExpectedRuntimeFactor() * int64(math.Floor(expectedRuntime.Minutes()/float64(num)))
+	unit.cachedValue += priority * unit.distro.GetTimeInQueueFactor() * int64(math.Floor(timeInQueue.Minutes()/float64(num)))
+	return unit.cachedValue
+}
+
+// StringSet provides simple tools for managing sets of strings.
+type StringSet map[string]struct{}
+
+// Add places the string in the set.
+func (s StringSet) Add(id string) { s[id] = struct{}{} }
+
+// Check returns true if the string is a member of the set.
+func (s StringSet) Check(id string) bool { _, ok := s[id]; return ok }
+
+// Visit returns true if the string is already a member of the
+// set. Otherwise it adds it to the set and returns false.
+func (s StringSet) Visit(id string) bool {
+	if s.Check(id) {
+		return true
+	}
+
+	s.Add(id)
+	return false
+}
+
+// TaskList implements sort.Interface on top of a slice of tasks. The
+// provided sorting, orders members of task groups, and then
+// prioritizes tasks by the number of dependencies, priority, and
+// expected duration.
+type TaskList []task.Task
+
+func (tl TaskList) Len() int      { return len(tl) }
+func (tl TaskList) Swap(i, j int) { tl[i], tl[j] = tl[j], tl[i] }
+func (tl TaskList) Less(i, j int) bool {
+	t1 := tl[i]
+	t2 := tl[j]
+
+	if t1.TaskGroupOrder != t2.TaskGroupOrder {
+		return t1.TaskGroupOrder < t2.TaskGroupOrder
+	}
+
+	if t1.NumDependents != t2.NumDependents {
+		return t1.NumDependents > t2.NumDependents
+	}
+
+	if t1.Priority != t2.Priority {
+		return t1.Priority > t2.Priority
+	}
+
+	return t1.FetchExpectedDuration() > t2.FetchExpectedDuration()
+}
+
+// TaskPlan provides a sortable interface on top of a slice of
+// schedulable units, with ordering of units provided by the
+// implementation of RankValue.
+type TaskPlan []*Unit
+
+func (tpl TaskPlan) Len() int           { return len(tpl) }
+func (tpl TaskPlan) Less(i, j int) bool { return tpl[i].RankValue() > tpl[j].RankValue() }
+func (tpl TaskPlan) Swap(i, j int)      { tpl[i], tpl[j] = tpl[j], tpl[i] }
+
+func (tpl TaskPlan) Keys() []string {
+	out := []string{}
+	for _, unit := range tpl {
+		out = append(out, unit.Keys()...)
+	}
+	return out
+}
+
+// PrepareTasksForPlanning takes a list of tasks for a distro and
+// returns a TaskPlan, grouping tasks into the appropriate units.
+func PrepareTasksForPlanning(distro *distro.Distro, tasks []task.Task) TaskPlan {
+	cache := UnitCache{}
+
+	for _, t := range tasks {
+		var unit *Unit
+		if t.TaskGroup != "" {
```

Comment:
If a task group and `ShouldGroupVersions`, it isn't grouped. What will the effect of this be?

Reply:
fixed.